src/createAction.js

```diff
@@ -10,35 +9,30 @@ export default function createAction(type, payloadCreator = identity, metaCreato
     'Expected payloadCreator to be a function, undefined or null'
   );
 
-  const finalPayloadCreator = isNull(payloadCreator)
+  const finalPayloadCreator = payloadCreator === null || payloadCreator === identity
     ? identity
-    : payloadCreator;
+    : (head, ...args) => head instanceof Error ?
+        head : payloadCreator(head, ...args);
+  
+  const hasMeta = isFunction(metaCreator);
 
   const actionCreator = (...args) => {
-    const hasError = args[0] instanceof Error;
-
-    const action = {
-      type
-    };
-
-    const payload = hasError ? args[0] : finalPayloadCreator(...args);
-    if (!isUndefined(payload)) {
+    const payload = finalPayloadCreator(...args);
+    const action = { type, error: payload instanceof Error };
+    
+    if (payload !== undefined) {
       action.payload = payload;
     }
-
-    if (hasError || payload instanceof Error) {
-      // Handle FSA errors where the payload is an Error object. Set error.
-      action.error = true;
-    }
-
-    if (isFunction(metaCreator)) {
+    
+    if (hasMeta) {
       action.meta = metaCreator(...args);
     }
 
     return action;
   };
 
-  actionCreator.toString = () => type.toString();
+  const typeStr = type.toString();
```

Comment:
It means that we don't have to call `type.toString` every time `actionCreator.toString` is called.